### ISSUE TITLE
Reduce size of Merkle proofs (we can derive pos for each path sibling)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#797]
Pretty sure we can avoid storing the pos for each node in the path in the Merkle proof.

For a given MMR size the pos are all known ahead of time.

I think if we know the pos of the node we care about and the (unpruned) size of the MMR then we have everything we need to determine the pos of each sibling in the path.

We would presumably do something like -

* for a given node pos -
  * find the parent and the sibling pos (which we know how to do, via family())
  * repeat up to the peak (staying inside the overall MMR based on unpruned size)

So as an alternative to storing the pos for each sibling in path - we can simply store the pos of the single leaf node and the total unpruned size of the MMR.

👉 Might be a good idea to try and get this in before testnet2 to avoid having to think about backward compatibility...

**Comment by tromp** [comment]
You're right that all left/right information in a Merkle tree path follows from the leaf index and MMR size.
E.g. the path from the root to leaf 8 (1000 in binary) is
RLLL in any MMR of size >= 13
RLL in a size 11 or 12 MMR
RL in a size 10 MMR
R in a size 9 MMR

**Comment by antiochp** [comment]
(MMR is still blowing my mind at times)

**Comment by ignopeverell** [comment]
@tromp you might find the set of functions following this interesting, along those same lines:

https://github.com/mimblewimble/grin/blob/master/core/src/core/pmmr.rs#L800

**Comment by tromp** [comment]
I think it may be possible to streamline the Merkle tree indexing.
My first inkling is to define 0-based leaf and node positions.
Then a function mapping a leaf to a node index would look like

    node = (leaf << 1) - ctpop64(leaf)

**Comment by tromp** [comment]
There's another inefficiency in Merkle proofs:

	/// The full list of peak hashes in the MMR
	pub peaks: Vec<Hash>,

There's no need to have the full list of peaks; all peaks smaller than the one our leaf is in can be replaced by a single hash.

**Comment by tromp** [comment]
peaks are hash together from smallest size (depicted on the right) to largest size (depicted on left)
so i would collapse all peaks on the right.
if there are 4 peaks, P0,P1,P2,P3 then the Merkle root is hash(P0, hash(P1,H))
where H = hash(P2,P3), to prove a leaf in P1 you'd just use H rather than P2 and P3.

**Comment by antiochp** [comment]
Does the order actually matter? I think (could be mistaken) that we are doing the reverse of this.
hash(P0,P1) then feeding that as H into hash(H, P2) etc.

Coinbase spends are presumably likely to be relatively young (miner sweeping them as soon as they mature) - so they are going to be on the right hand side of the MMR (most of the time).

So feels like we should maybe prehash the larger number of peaks on the left?

**Comment by antiochp** [comment]
Hmm actually this is pretty counter-intuitive. Those on the left are going to grow a _lot_ slower than those on the right. And those on the right are going to increase and decrease over time relatively quickly.
Coinbase matures after 1000 blocks, so we have peaks to the right for -
512
256
128
64
32
16
8
4
2
leaves that come and go.

**Comment by tromp** [comment]
I see i listed the peaks in the wrong left-to-right order. Fixed now.

**Comment by antiochp** [comment]
We could get smart and selectively hash left-to-right or right-to-left to give us the smallest Merkle proof?

**Comment by tromp** [comment]
You can see in https://github.com/opentimestamps/opentimestamps-server/blob/master/doc/merkle-mountain-range.md how the peaks are hashed together, from smaller on right first.

**Comment by antiochp** [comment]
I don't actually see where it illustrates that?
Not disagreeing but pretty sure we did the opposite of that in our MMR impl.

**Comment by tromp** [comment]
While the small peaks on the right have the most recent txos, you must realize that with 20 peaks in total,
only 0.1% of all txo are in the rightmost 10 peaks. So you can still expect a lot of spending from the 99.9% of txo in the leftmost 10 peaks.

**Comment by tromp** [comment]
It illustrates it in this diagram
```
      /\
     /  \
    /    \
   /\     \
  /  \    /\
 /\  /\  /\ \
/\/\/\/\/\/\/\
```
which would otherwise look like
```
        /\
       / /\
      /\
     / /\
    / /\/\
   /\     
  /  \    
 /\  /\  
/\/\/\/\
```

**Comment by antiochp** [comment]
But where does it actually say we hash the peaks right to left? That diagram (the former) would work equally well for hashing the peaks in either direction?

**Comment by tromp** [comment]
The hash order is clearly implied in the picture; each node with left and right children represents the hash of those children.

**Comment by tromp** [comment]
I'm rather confused by these lines:

	///  * inclusion of the node beneath a peak (via the Merkle path/branch of siblings)
	///  * inclusion of the peak in the "bag of peaks" beneath the root
	pub fn verify(&self) -> bool {
		// if we have no further elements in the path
		// then this proof verifies successfully if our node is
		// one of the peaks
		// and the peaks themselves hash to give the root
		if self.path.len() == 0 {
			if !self.peaks.contains(&self.node) {

This tests whether node is one of the peaks in the peaks vector, not whether it is a node below a peak?!

**Comment by antiochp** [comment]
Oh ok. Point taken. I was not actually reading that diagram as there being an intermediate "node" connecting the 2nd and 3rd trees. And that the layout was simply an artifact of it being a ascii diagram... 
I was reading it as a single root with all 3 peaks as direct children (and the actual hashing impl left as an exercise for the reader).

**Comment by antiochp** [comment]
`proof.verify()` is recursive as we traverse up the MMR - the node _will_ be one of the peaks in the final recursion step.

**Comment by tromp** [comment]
oh, i should've read the comments in the middle! deleted my embaressing question:-(

**Comment by antiochp** [comment]
Just to confuse matters, this is an alternative visualization with the roots hashed the other way -
https://github.com/proofchains/python-proofmarshal/blob/master/proofmarshal/mmr.py#L60-L86

**Comment by tromp** [comment]
That alternative visualization is equivalent to my second. I just drew mine in a misleading way. Fixed now.

**Comment by tromp** [comment]
I don't see where they do that either. That would require rehashing every node in the MMR whenever you add one element. Which would defeat the whole purpose...

**Comment by tromp** [comment]
I'd like to propose a small overhaul of the MMR model, with slightly more efficient primitives.
It will store the MMR in a single vector of nodes, where the peaks would just be implicit.
The nodes are numbered from 0 rather than 1. After adding 4 leaves, we have
```
   6
  / \  
 2   5
/ \ / \ 
0 1 3 4 
```
The current size, and next position in the vector, is 7. And we have a single peak of that size.
Let the peakmap be a bitmap of occupied peaks. Bit n is 1 if we have a peak of size 2^{n+1}-1.
Here the peakmap is 100. After adding 3 more leaves, we'll have 3 peaks of sizes 7, 3, and 1
and a peakmap of 111.

When adding a leaf, its path from the root will follow the current peakmap.
Adding a fifth leaf to the above tree, will put it at 100, or right-left-left.
This also immediately tells us that no additional parent nodes will be added.
```
   6
  / \  
 2   5
/ \ / \ 
0 1 3 4 7
```
The next leaf goes at 101, or right-left-right.
For a path ending in k 1s (rights), k parent nodes will be added, in this case k=1:
```
   6
  / \  
 2   5   9
/ \ / \ / \
0 1 3 4 7 8
```
Adding one more leaf at 110 gives us
```
   6
  / \  
 2   5   9
/ \ / \ / \  /
0 1 3 4 7 8 10
```
The offsets in the single vector for the 3 peaks, from small to large, are at 10, 9, and 6,
differing by the 2^i-1 sizes, so we can easily iterate over them.
And now the next leaf goes at 111, with k=3 parents added:
```
       14
      /  \
     /    \
    /      \
   6       13
  / \     /  \
 2   5   9    12
/ \ / \ / \  /  \
0 1 3 4 7 8 10  11
```
bringing us to peakmap 1000.

In this tree, finding the height of and path to a node is a simple matter of repeatedly subtracting sizes 2^i-1.
For node 9, we get 9-7=2, 2<3, 2-1=1, so the peakmap at time of addition of 9 was 101,
and its height is the remainder, 1.

**Comment by tromp** [comment]
I also propose a slight overhaul of Merkle proofs.
Say we want to prove inclusion of leaf 8 in MMR
```
       14
      /  \
     /    \
    /      \
   6       13
  / \     /  \
 2   5   9    12      17
/ \ / \ / \  /  \    /  \
0 1 3 4 7 8 10  11  15  16  18
```
Then we'd first "bag the peaks", from smallest (rightmost) to largest, as in
```
              .
             / \
            /   \ 
           /     \ 
          /       \
         /         \
       14           \
      /  \           \
     /    \           \
    /      \           \
   6       13           .
  / \     /  \         / \
 2   5   9    12      17  \
/ \ / \ / \  /  \    /  \  \
0 1 3 4 7 8 10  11  15  16  18
```
to get a single root, and then provide the usual Merkle tree proof, ignoring the location of peaks.
So the proofs for leaves up to 11 would not need to list the separate peaks at 17 and 18.

**Comment by antiochp** [comment]
From @tromp in gitter - 

> merkle proof for 8 in my final MMR would be 8, 8data, hashes at [8,7,12,6,19] where 19 is the parent of 17 and 18

If we hash peaks right to left, we can just treat "intermediate nodes above the peaks" the same way as other non-leaf nodes for the purpose of building the Merkle proof.

**Comment by tromp** [comment]
what index would you use for hashing at those bagging nodes to "treat them the same way" ?

**Comment by antiochp** [comment]
> the verify routine should really take a data item, a merkle root, and a merkle proof
where merkle proof has the leaf position, mmr size, and vector of sibling hashes only

**Comment by antiochp** [comment]
@tromp - it has to be "size" for each of them right? This is where they differ slightly from regular nodes.

**Comment by tromp** [comment]
yes, exactly. i thought mmr size is the most natural choice for index at those bagging nodes

**Comment by antiochp** [comment]
This has the added benefit of including the size in the data being hashed at the root, which we don't have currently.

**Comment by tromp** [comment]
agreed, it's good to commit to the mmr size somewhere

**Comment by ignopeverell** [comment]
The rightmost peak always indirectly commit to the size, just due to its own position.

**Comment by antiochp** [comment]
Oh yeah of course. Was forgetting that the rightmost pos _is_ the unpruned size.

**Comment by tromp** [comment]
i'll try to prepare a PR for this...

**Comment by sergorl** [comment]
I'm sure there is enough to store only siblings in field path, because I can recursively calculate next tree node (which is peak) in the next stage of height, using only its sibling in the current level - so I will get finished calculated root. Here is the scheme-picture to this situation: [png](https://github.com/sergorl/docs/blob/master/sibling.png)
Then I just compare calculated root and original root. Isn't it?

**Comment by tromp** [comment]
Yes, all we need in a proof is

MMR size
leaf position
leaf hash
sibling hashes on path to root

**Comment by ignopeverell** [comment]
So given that we're about to release testnet2, and with this possible improvement, how about we change the MMR root function to hash from right instead? Then the 2 approaches would be compatible and we can just have hash_with do a +/-1. And we can compare them side-by-side.

**Comment by tromp** [comment]
During a discussion with jasper and yeastplume, I realized that Merkle proofs all the way up to the TXO root are not that useful, since the top sibling changes every block. This makes Merkle proofs very maintenance intensive and fast expiring.
I propose that they instead only prove inclusion under a peak. The verifier should be able to verify that said peak is actually present.

**Comment by antiochp** [comment]
The root may change every block but a verifier will have all block headers, one of which will contain that root?

**Comment by tromp** [comment]
Following igno's suggestion, i'll first make a PR for just the consensus changes,
and then continue the rewrite later after testnet2 launch.

**Comment by tromp** [comment]
Antioch, besides coinbase maturity proofs, would merkle proofs to the root have any use?
Would you agree that for spending very old outputs, proofs of inclusion under a current peak are more useful?

**Comment by antiochp** [comment]
Actually I take that back - rewind is not going to work for really old outputs. If peaks are getting rolled up into higher peaks then we cannot rollback to a point prior to that.

So maybe a proof against a peak of a large tree would potentially be useful here.

I still think a validator having all block headers (and therefore all roots for each block height) means we can do the same thing with a proof against the root.

**Comment by tromp** [comment]
yes, under-peak proofs will expire with exponentially decreasing frequency.
that's rare enough that they can be easily reconstructed if necessary.

once we have under-peak proofs (which are needed for spending very old outputs),
then under-root proofs are best implemented as an under-peak proof wrapped in a peak-under-root proof.

**Comment by ignopeverell** [comment]
Also note than a proof against the root of an old tree, in an older block header, is more secure than one against a peak in a recent tree, due to the possibility of reorgs (which could invalidate parts of the proof).

**Comment by antiochp** [comment]
Resolved (for testnet3) in #1119.
